pallets/proofs-dealer/README.md

```diff
@@ -0,0 +1,120 @@
+
+The reason for this distinction can be found in the [PR](https://github.com/paritytech/polkadot-sdk/pull/1781) that introduced the poll hook. During Multi Block Migrations, `on_poll` is not called, therefore the challenges ticker wouldn't be incremented. That is a desirable behaviour, since during Multi Block Migrations, Providers wouldn't have the opportunity to submit proofs. By not incrementing the challenges ticker, we ensure that no Provider is slashed during Multi Block Migrations.
+
+![Substrate block lifecycle](./resources/blockLifecycle.png)
+
+### Random Challenge Distribution and Enforcement
+
+A random seed is generated and stored at every tick. Each Provider has a "challenge period" of a number of ticks, that determines which ticks it has to look for, to know the challenges it would have to respond to. In the image above, we can see 3 Providers of different sizes, which therefore have different challenge periods. The challenge period is calculated as `stake / StakeToChallengePeriod`, where `StakeToChallengePeriod` is a configurable constant. In the example, the challenge period for the smallest Provider is 5 ticks, having to answer to challenges in blocks `i-5` and `i`. In a real world scenario, periods would be much longer, but for the sake of the diagram, it is kept short.
+
+A Provider has a number of ticks as tolerance to respond to challenges. This tolerance is configured by the `ChallengesTicksTolerance` constant, and is common to all Providers, regardless of their size. In the example diagram, the tolerance is set to 2 ticks, so our small Provider would have submitted a proof for challenges in blocks `i-5`, either in block `i-3` or `i-4`.
+
+In the image, it can also be observed that the current block is `i`, which contains the new challenge seed that our small Provider has to use to generate the challenges it is supposed to respond to. The seed is stored in the `TickToChallengesSeed` storage map, which is accessed with a tick number as key. The `TickToChallengesSeed` stores a `ChallengeHistoryLength` number of seeds, and the oldest seed is removed at every tick, before adding the new seed. With the seed, the Provider will generate a number of challenges set by the `RandomChallengesPerBlock` constant, which are made of the concatenation of the seed, the Provider's ID and an index, all SCALE-encoded, and later hashed. This is done in the following way:
+
+```rust
+for i in 0..RandomChallengesPerBlock {
+    let challenge = T::MerkleTrieHashing::hash(
+        &[
+            seed.encode().as_ref(),
+            provider_id.encode().as_ref(),
+            i.encode().as_ref(),
+        ]
+        .concat(),
+    );
+}
+```
+
+At the end of the tolerance period, if the Provider has not submitted a proof for the challenges it is supposed to respond to, it is marked as "slashable". How that slashing is implemented is out of scope for this pallet.
```

Comment:
Comment unrelated to docs but regarding this logic: have we thought about how we could handle when a provider should be slashed multiple times? Since we only mark it as slashable, if between two non-delivered proof the slashing system doesn't get to that provided, it could avoid being double-slashed as it should. Maybe it makes sense to keep a counter of not provided proofs instead?

Reply:
Yes, I've given it a thought and it can be achieved by something like you suggested, of keeping a counter. However I don't know if it is worth it to continue to complicate an already very complex pallet, given that this could also be solved by adding an offchain worker that just sends the slash transaction whenever there are Providers marked as slashable. After all, StorageHub collators would be incentivised to enforce slashing if that slashing goes to a treasury, which in turn goes to pay the collators.